Add prometheus metrics

jdreichmann opened the pull request:
Continuation of #71

The following metrics are available for scraping with prometheus:

- `bridge_rooms_total`: labeled by protocol and type (dm, group)

- `bridge_puppets_total`: labeled by protocol, how many puppets
exist (aka remote accounts matrix users have linked)

- `bridge_remote_users_total`: labeled by protocol, how many
remote users (sometimes called "ghosts") exist

- `bridge_messages_total`: labeled by protocol and type, so
`count(bridge_messages_total) by (protocol, type)` is possible

- `bridge_matrix_events_total`: labeled by protocol and type,
the total amount of matrix events processed since startup

- `bridge_matrix_event_errors_total`: labeled by protocol,
how many matrix events resulted in an exception during bridging

- `bridge_matrix_event_seconds`: histogram by protocol and type
of the time spent processing a matrix event to be bridged

- `bridge_remote_update_seconds`: histogram by protocol and type
of the time spent bridging remote events into matrix

Labeling by the protocol id enables you to have a single
scrape config in prometheus with many mx-puppet-* bridges
as targets, and easily visualizing global bridge activity
with f.ex. grafana.

For `bride_connected` to work, the implementing bridge needs
to call `trackConnectionStatus` on connect/disconnect.

tulir committed:
Add basic prometheus setup

jdreichmann committed:
feat: collect and serve prometheus metrics

The following metrics are available for scraping with prometheus:

- `bridge_rooms_total`: labeled by protocol and type (dm, group)

- `bridge_puppets_total`: labeled by protocol, how many puppets
exist (aka remote accounts matrix users have linked)

- `bridge_remote_users_total`: labeled by protocol, how many
remote users (sometimes called "ghosts") exist

- `bridge_messages_total`: labeled by protocol and type, so
`count(bridge_messages_total) by (protocol, type)` is possible

- `bridge_matrix_events_total`: labeled by protocol and type,
the total amount of matrix events processed since startup

- `bridge_matrix_event_errors_total`: labeled by protocol,
how many matrix events resulted in an exception during bridging

- `bridge_matrix_event_seconds`: histogram by protocol and type
of the time spent processing a matrix event to be bridged

- `bridge_remote_update_seconds`: histogram by protocol and type
of the time spent bridging remote events into matrix

- `bridge_connected`: currently active connections to the
remote network, labeled by protocol

Labeling by the protocol id enables you to have a single
scrape config in prometheus with many mx-puppet-* bridges
as targets, and easily visualizing global bridge activity
with f.ex. grafana.

For `bride_connected` to work, the implementing bridge needs
to call `trackConnectionStatus` on connect/disconnect.

Sorunome commented on src/roomsyncroniser.ts:
```diff
@@ -51,6 +51,11 @@ export class RoomSyncroniser {
+		if (this.config.metrics.enabled) {
+			const rooms = roomStore.getAll();
+			this.bridge.bridgeRoomMetric.set({type: "dm", protocol: this.bridge.protocol.id}, rooms.filter(room => room.isDirect).length))
+			this.bridge.bridgeRoomMetric.set({type: "group", protocol: this.bridge.protocol.id}, rooms.filter(room => room.isDirect).length))
```
you meant to filter `!room.isDirect`? 😉

jdreichmann replied:
ah, yes... i was missing coffee :)

jdreichmann replied:
GitHub doesn't like force-pushes, it seems - changed in the newest version

Sorunome commented on src/roomsyncroniser.ts:
```diff
@@ -51,6 +51,11 @@ export class RoomSyncroniser {
 		this.roomStore = this.bridge.roomStore;
 		this.mxidLock = new Lock(MXID_LOOKUP_LOCK_TIMEOUT);
+		if (this.config.metrics.enabled) {
+			const rooms = roomStore.getAll();
```
roomStore.getAll is a promise and promises can't be `await`'ed in a constructor. so, you'd need to do something like

```
roomStore.getAll().then((rooms) => {
  this.bridge.blahblahblah
});
```

jdreichmann commented on src/puppetbridge.ts:
```diff
@@ -513,7 +558,12 @@ export class PuppetBridge extends EventEmitter {
 		const room = await this.namespaceHandler.createRoom(roomData);
-		if (!room || room.isDirect) {
+		if (room) {
+			this.metrics.room.inc({type: room.isDirect ? "dm" : "group", protocol: this.protocol?.id});
```
This only increments for group-chats, never for DMs - @Sorunome am i checking the correct thing here? is `room.isDirect` guaranteed always set correctly here?

jdreichmann committed:
Add database latency metrics

This commit defines a `bridge_database_query_seconds` histogram
which records the duration of database queries.

Sorunome commented on src/db/emotestore.ts:
```diff
@@ -36,44 +37,51 @@ export class DbEmoteStore {
 	}
 
 	public async get(puppetId: number, roomId: string | null, emoteId: string): Promise<IEmoteStoreEntry | null> {
+		const stopTimer = this.db.latency.startTimer(this.labels("select"));
```
Are these things any significant performance hits if metrics are disabled?

jdreichmann replied:
I haven't noticed a significant performance hit but i also didn't run it through a profiler. The underlying function uses `process.hrtime` and calculates the diff, so there might be a tiny memory penalty if a lot of timers run at the same time. I couldn't get more metrics on it, because, well...

I suppose it's fine, i haven't noticed any slowdown

Sorunome commented on src/db/userstore.ts:
```diff
@@ -37,7 +40,30 @@ export class DbUserStore {
 
+	public async getAll() {
+		const stopTimer = this.db.latency.startTimer(this.labels("select_all"));
+		const results: object[] = [];
```
Please use an appropriate type here, not just `object[]` (`IUserStoreEntry[]`?)

Sorunome commented on src/puppetbridge.ts:
```diff
@@ -211,6 +229,37 @@ export class PuppetBridge extends EventEmitter {
 
+		if (this.config.metrics.enabled) {
+			prometheus.collectDefaultMetrics();
+			const metricsServer = express();
```
why don't we re-use our existing express instance, so that we only have to have one single http server running?

jdreichmann replied:
As far as i could see, there is no simple way to listen on different ports with the same express instance - it seems to be relatively common to have metrics on a seperate port from the "main service", but i haven't looked into the common way of archieving that.

It also makes routing and controlling both application and metrics traffic possible on layer 3, by just applying f.ex. different firewall zones, potentially improving complexity of reverse-proxy-setups.

jdreichmann replied:
BUMP

Sorunome commented on src/puppetbridge.ts:
```diff
@@ -513,7 +563,12 @@ export class PuppetBridge extends EventEmitter {
 		// check if this is a valid room at all
 		const room = await this.namespaceHandler.createRoom(roomData);
-		if (!room || room.isDirect) {
+		if (room) {
```
```suggestion
		if (!room) {
			return;
		}
		this.metrics.room.inc({type: room.isDirect ? "dm" : "group", protocol: this.protocol?.id});
```

jdreichmann replied:
see https://github.com/Sorunome/mx-puppet-bridge/pull/72/files#diff-4cd484140d4c36875ceef0af75412976a1090961b49b5137435995661f73cb44R568 - unfortunately, githab has no "yes this is in the commit now"-detection

Sorunome commented on src/puppetbridge.ts:
```diff
@@ -671,10 +727,25 @@ export class PuppetBridge extends EventEmitter {
 	 */
-	public async sendStatusMessage(puppetId: number | IRemoteRoom, msg: string) {
+	public async sendStatusMessage(puppetId: number | IRemoteRoom, msg: string, isConnected: boolean | null = null) {
+		if (isConnected !== null && typeof puppetId === "number") {
```
```suggestion
		if (isConnected !== null) {
			this.trackConnectionStatus(typeof puppetId === "number" ? puppetId : puppetId.puppetId, isConnected);
```

jdreichmann replied:
see https://github.com/Sorunome/mx-puppet-bridge/pull/72/files#diff-4cd484140d4c36875ceef0af75412976a1090961b49b5137435995661f73cb44R744 - added it to the commit

Sorunome commented on src/roomsyncroniser.ts:
```diff
@@ -51,6 +51,28 @@ export class RoomSyncroniser {
+			};
+			this.roomStore.getAll()
+				.catch((err) => log.error("could not get room store"))
+				.then(roomMetricsInit)
```
You have a race condition here. In the `puppetbridge.ts` initalizing you set `this.bridge.metrics.room` only **after** you created the roomSync object. So you should move the stuff in puppetbridge.ts up a bit, probably

jdreichmann replied:
Moved that part where `this.bridge.metrics` gets populated into the constructor of the PuppetBridge - see https://github.com/Sorunome/mx-puppet-bridge/pull/72/files#diff-4cd484140d4c36875ceef0af75412976a1090961b49b5137435995661f73cb44R164

Sorunome reviewed (approved):
Awesome, thank you so much!
